File: Heaps/Top-K-Frequent-Elements/top-k.cpp
#include <vector>
#include <queue>
#include <unordered_map>
using namespace std;

class Solution {
public:
    typedef pair<int,int> pip;
    vector<int> topKFrequent(vector<int>& nums, int k) {
        unordered_map<int,int>mp;
        int n = nums.size();
        for(int i=0;i<n;i++){
            mp[nums[i]]++;
        }
        priority_queue<pip,vector<pip>,greater<pip>>pq;
        for(auto p: mp){
            int freq = p.second;
            int ele = p.first;
            pq.push({freq,ele});
            if(pq.size()>k)pq.pop();
        }
        vector<int>ans;
        while(!pq.empty()){
            ans.push_back(pq.top().second);
            pq.pop();
        }
        return ans;
    }
};

File: Heaps/Top-K-Frequent-Elements/top-k.py
import heapq
from typing import List

class Solution:
    def topKFrequent(self, nums: List[int], k: int) -> List[int]:
        mp = {}
        n = len(nums)
        for i in range(n):
            mp[nums[i]] = mp.get(nums[i],0) + 1
        heap = []
        for key,value in mp.items():
            heap.append((value,key))
            heapq.heapify(heap)
            if len(heap) > k: heapq.heappop(heap)
        ans = []
        while len(heap):
            ans.append(heap[-1][1])
            heap.pop()
        return ans